### PR TITLE
Confluence storage and document plugin

### PR DESCRIPTION
**Pull request by Cinojose**
An attempt to create a confluence plugin for both storage and document. 

The plugin right now supports only cloud hosted confluence and can be extended down the road. Logs and sheets are not supported as confluence has no such concept. it make use of the _atlassian-python-api_ for the confluence api calls. 

I can create some screenshots on how it looks like, if there is a need. Please review.

### CODE REVIEW

File: src/dispatch/plugins/dispatch_confluence/plugin.py [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+
+    def __init__(self):
+        self.configuration_schema = ConfluenceConfiguration
+
+    def create_file(
+        self, drive_id: str, name: str, participants: List[str] = [], file_type: str = "folder"
+    ):
+        """Creates a new Home page for the incident documents.."""
+        try:
+            print(f"{drive_id} {name} {participants} {file_type}")
```
Comment (mvilanova) [inline src/dispatch/plugins/dispatch_confluence/plugin.py:57]:
```suggestion
```

File: src/dispatch/plugins/dispatch_confluence/plugin.py [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+                full_width=False,
+            )
+            return {
+                "weblink": f"{self.configuration.api_url}wiki/spaces/{drive_id}/pages/{page_details['id']}/{name}",
+                "id": page_details["id"],
+                "name": name,
+                "description": "",
+            }
+        except Exception as e:
+            print(f"Exception happened while creating page: {e}")
```
Comment (mvilanova) [inline src/dispatch/plugins/dispatch_confluence/plugin.py:87]: Let's use our logger instead. Example here: https://github.com/Netflix/dispatch/blob/4e8ad0b48857f1bc78122a58b1365ff1d6687a01/src/dispatch/common/utils/cli.py#L9

File: src/dispatch/plugins/dispatch_confluence/plugin.py [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+
+    def copy_file(self, folder_id: str, file_id: str, name: str):
```
Comment (mvilanova) [inline src/dispatch/plugins/dispatch_confluence/plugin.py:89]: Missing typing

File: src/dispatch/plugins/dispatch_confluence/plugin.py [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+        # TODO : This is the function that is responsible for making the incident documents.
+        try:
+            confluence_client = Confluence(
+                url=self.configuration.api_url,
+                username=self.configuration.username,
+                password=self.configuration.password.get_secret_value(),
+                cloud=self.configuration.hosting_type,
+            )
+            print(f"Copy_file function with args {folder_id}, {file_id}, {name}")
```
Comment (mvilanova) [inline src/dispatch/plugins/dispatch_confluence/plugin.py:98]: Remove or use logger.

File: src/dispatch/plugins/dispatch_confluence/plugin.py [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+            if self.configuration.parent_id:
+                """TODO: Find and fix why the page is not created under the parent_id, folder_id"""
+                self.move_file_confluence(page_id_to_move=page_details["id"], parent_id=folder_id)
+            return {
+                "weblink": f"{self.configuration.api_url}wiki/spaces/{folder_id}/pages/{page_details['id']}/{name}",
+                "id": page_details["id"],
+                "name": name,
+            }
+        except Exception as e:
+            print(f"Exception happened while creating page: {e}")
```
Comment (mvilanova) [inline src/dispatch/plugins/dispatch_confluence/plugin.py:121]: Use logger.

File: src/dispatch/plugins/dispatch_confluence/plugin.py [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+
+    def move_file(self, new_folder_id, file_id, **kwargs):
```
Comment (mvilanova) [inline src/dispatch/plugins/dispatch_confluence/plugin.py:123]: Missing typing.

File: src/dispatch/plugins/dispatch_confluence/plugin.py [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+        """Moves a file from one place to another. Not used in the plugin,
+        keeping the body as the interface is needed to avoid exceptions."""
+        return {}
+
+    def move_file_confluence(self, page_id_to_move, parent_id):
```
Comment (mvilanova) [inline src/dispatch/plugins/dispatch_confluence/plugin.py:128]: Missing typing.

File: src/dispatch/plugins/dispatch_confluence/plugin.py [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+        try:
+            url = f"{self.configuration.api_url}wiki/rest/api/content/{page_id_to_move}/move/append/{parent_id}"
+            auth = HTTPBasicAuth(
+                self.configuration.username, self.configuration.password.get_secret_value()
+            )
+            headers = {"Accept": "application/json"}
+            response = requests.request("PUT", url, headers=headers, auth=auth)
+            return response
+        except Exception as e:
+            print(f"Exception happened while moving page: {e}")
```
Comment (mvilanova) [inline src/dispatch/plugins/dispatch_confluence/plugin.py:138]: Use logger.

**Review by mvilanova — CHANGES_REQUESTED** [review]
Looks good so far. Just minor stuff. I think there's an opportunity to have all the atlassian-related plugins (Jira and Confluence) under the same directory now and have all of them leverage (and extend if necessary) the config stuff such as HostingType and API related things. We could tackle it in a separate PR.

File: setup.py [hunk 9]
```diff
@@ -423,6 +423,8 @@ def run(self):
             "slack_contact = dispatch.plugins.dispatch_slack.plugin:SlackContactPlugin",
             "slack_conversation = dispatch.plugins.dispatch_slack.plugin:SlackConversationPlugin",
             "zoom_conference = dispatch.plugins.dispatch_zoom.plugin:ZoomConferencePlugin",
+            "dispatch_confluence = dispatch.plugins.dispatch_confluence.plugin:ConfluencePagePlugin",
+            "dispatch_confluence_document = dispatch.plugins.dispatch_confluence.docs.plugin:ConfluencePageDocPlugin",
```
[the remark this answers is not in the document]
Comment (Cinojose) [reply setup.py]: Done

File: src/dispatch/incident/flows.py [hunk 10]
```diff
@@ -360,17 +360,17 @@ def create_incident_documents(incident: Incident, db_session: SessionLocal):
             sheet = plugin.instance.create_file(
                 incident.storage.resource_id, incident_sheet_name, file_type="sheet"
             )
+        if sheet:
```
[the remark this answers is not in the document]
Comment (Cinojose) [reply src/dispatch/incident/flows.py:364]: Done

File: src/dispatch/plugins/dispatch_confluence/config.py [hunk 11]
```diff
@@ -0,0 +1,28 @@
+from pydantic import Field, SecretStr, HttpUrl
+
+from enum import Enum
+from dispatch.config import BaseConfigurationModel
+
+
+class HostingType(str, Enum):
+    """Type of Confluence deployment."""
```
[the remark this answers is not in the document]
Comment (Cinojose) [reply src/dispatch/plugins/dispatch_confluence/config.py]: Done

File: src/dispatch/plugins/dispatch_confluence/config.py [hunk 12]
```diff
@@ -0,0 +1,28 @@
...
+
+    cloud = "cloud"
+    server = "server"
+
+
+class ConfluenceConfigurationBase(BaseConfigurationModel):
+    """Confluene configuration description."""
```
[the remark this answers is not in the document]
Comment (Cinojose) [reply src/dispatch/plugins/dispatch_confluence/config.py]: Done

File: src/dispatch/plugins/dispatch_confluence/plugin.py [hunk 13]
```diff
@@ -0,0 +1,138 @@
...
+from atlassian import Confluence
+import requests
+from requests.auth import HTTPBasicAuth
+import json
+from typing import List
+
+
+# TODO : Use the common config from the root directory.
+class ConfluenceConfiguration(ConfluenceConfigurationBase):
+    """Confluene configuration description."""
```
[the remark this answers is not in the document]
Comment (Cinojose) [reply src/dispatch/plugins/dispatch_confluence/plugin.py]: Done

File: src/dispatch/plugins/dispatch_confluence/plugin.py (hunk 4 above)
Comment (Cinojose) [reply src/dispatch/plugins/dispatch_confluence/plugin.py]: Updated to logger.

File: src/dispatch/plugins/dispatch_confluence/plugin.py (hunk 5 above)
Comment (Cinojose) [reply src/dispatch/plugins/dispatch_confluence/plugin.py]: Done

File: src/dispatch/plugins/dispatch_confluence/plugin.py (hunk 6 above)
Comment (Cinojose) [reply src/dispatch/plugins/dispatch_confluence/plugin.py]: Done

File: src/dispatch/plugins/dispatch_confluence/plugin.py (hunk 7 above)
Comment (Cinojose) [reply src/dispatch/plugins/dispatch_confluence/plugin.py]: Updated.

File: src/dispatch/plugins/dispatch_confluence/plugin.py (hunk 8 above)
Comment (Cinojose) [reply src/dispatch/plugins/dispatch_confluence/plugin.py]: Updated.

File: src/dispatch/plugins/dispatch_confluence/plugin.py (hunk 3 above)
Comment (Cinojose) [reply src/dispatch/plugins/dispatch_confluence/plugin.py]: Fixed.

File: src/dispatch/plugins/dispatch_confluence/plugin.py (hunk 2 above)
Comment (Cinojose) [reply src/dispatch/plugins/dispatch_confluence/plugin.py]: Fixed.

File: src/dispatch/plugins/dispatch_confluence/plugin.py (hunk 1 above)
Comment (Cinojose) [reply src/dispatch/plugins/dispatch_confluence/plugin.py]: Removed.

File: src/dispatch/plugins/dispatch_confluence/plugin.py [hunk 14]
```diff
@@ -0,0 +1,138 @@
...
+        title="Readonly",
+        default=False,
+        description="The incident document will be marked as readonly on incident close. Participants will still be able to interact with the document but any other viewers will not.",
+    )
+
+
+class ConfluencePagePlugin(StoragePlugin):
+    title = "Confluence Plugin - Store your incident details"
+    slug = "confluence"
+    description = "Conflunce plugin to create Incident documents"
```
[the remark this answers is not in the document]
Comment (Cinojose) [reply src/dispatch/plugins/dispatch_confluence/plugin.py]: Updated.

**Comment by mvilanova** [comment]
Thanks for the PR!

**Pull request merged**